base/strings/search.jl

```diff
@@ -23,176 +44,190 @@ julia> search("Hello to the world", "z")
             if c != d
                 matched = false
                 break
             end
         end
-        if matched
-            return i
-        end
+        matched && return i
         i = ii
     end
 end
 
-_searchindex(s, t::Char, i) = search(s, t, i)
-
 function _search_bloom_mask(c)
     UInt64(1) << (c & 63)
 end
 
-_nthbyte(s::String, i) = codeunit(s, i)
-_nthbyte(a::ByteArray, i) = a[i]
+_nthbyte(s::String, i) = @inbounds codeunit(s, i)
+_nthbyte(a::ByteArray, i) = @inbounds a[i]
 
 function _searchindex(s::Union{String,ByteArray}, t::Union{String,ByteArray}, i)
-    n = sizeof(t)
-    m = sizeof(s)
+    # assumes that caller has checked if i is a valid index
+    isempty(t) && return i
 
-    if n == 0
-        return 1 <= i <= m+1 ? max(1, i) : 0
-    elseif m == 0
-        return 0
-    elseif n == 1
-        return search(s, _nthbyte(t,1), i)
-    end
+    m = sizeof(s)
+    n = sizeof(t)
 
+    n == 0 && return i
+    n == 1 && return search(s, _nthbyte(t, 1), i)
     w = m - n
-    if w < 0 || i - 1 > w
-        return 0
-    end
+    w < i - 1 && return 0
 
     bloom_mask = UInt64(0)
     skip = n - 1
-    tlast = _nthbyte(t,n)
+    tlast = _nthbyte(t, n)
     for j in 1:n
-        bloom_mask |= _search_bloom_mask(_nthbyte(t,j))
-        if _nthbyte(t,j) == tlast && j < n
+        bloom_mask |= _search_bloom_mask(_nthbyte(t, j))
+        if _nthbyte(t, j) == tlast && j < n
             skip = n - j - 1
         end
     end
 
     i -= 1
     while i <= w
-        if _nthbyte(s,i+n) == tlast
+        if _nthbyte(s, i + n) == tlast
             # check candidate
             j = 0
             while j < n - 1
-                if _nthbyte(s,i+j+1) != _nthbyte(t,j+1)
+                if _nthbyte(s, i + j + 1) != _nthbyte(t, j + 1)
                     break
                 end
                 j += 1
             end
 
             # match found
             if j == n - 1
-                return i+1
+                return i + 1
             end
 
             # no match, try to rule out the next character
-            if i < w && bloom_mask & _search_bloom_mask(_nthbyte(s,i+n+1)) == 0
+            if i < w && bloom_mask & _search_bloom_mask(_nthbyte(s, i + n + 1)) == 0
                 i += n
             else
                 i += skip
             end
         elseif i < w
-            if bloom_mask & _search_bloom_mask(_nthbyte(s,i+n+1)) == 0
+            if bloom_mask & _search_bloom_mask(_nthbyte(s, i + n + 1)) == 0
                 i += n
             end
         end
         i += 1
     end
-
     0
 end
 
-searchindex(s::ByteArray, t::ByteArray, i) = _searchindex(s,t,i)
-
 """
-    searchindex(s::AbstractString, substring, [start::Integer])
+    searchindex(s, t, [start::Integer])
 
-Similar to [`search`](@ref), but return only the start index at which
-the substring is found, or `0` if it is not.
+Similar to [`search`](@ref), but return only the start index at which the second argument
```

Comment:
"substring" was clearer, let's not make the docs worse because of that `UInt8` thing.